Objects/listobject.c

```diff
@@ -76,6 +76,32 @@ list_resize(PyListObject *self, Py_ssize_t newsize)
     return 0;
 }
 
+static int
+list_preallocate_exact(PyListObject *self, Py_ssize_t size)
+{
+    PyObject **items;
+    size_t allocated, num_allocated_bytes;
+
+    allocated = (size_t)size;
+    if (allocated > (size_t)PY_SSIZE_T_MAX / sizeof(PyObject *)) {
+        PyErr_NoMemory();
+        return -1;
+    }
+
+    if (size == 0) {
+        allocated = 0;
+    }
+    num_allocated_bytes = allocated * sizeof(PyObject *);
+    items = (PyObject **)PyMem_Malloc(num_allocated_bytes);
+    if (items == NULL) {
+        PyErr_NoMemory();
+        return -1;
+    }
+    self->ob_item = items;
```

Comment:
It is not guarantied that ob_item is NULL here. New items can be added in destructors called in `_list_clear()`.

Reply:
In b678442 I modified the enter condition for the new preallocation so it goes directly to the `list_extend` path if there are objects after calling `_list_clear(self)`:

```C
if (_PyObject_HasLen(iterable) && self->ob_item == NULL) {
```

Reply:
Well, it wouldn't hurt to add assert(self->ob_item == NULL) at this function start. I see that as documentation of the function contract ;-)

Reply:
`PyObject_Size()` can call the Python code and release GIL. This can lead to adding items to the list.

Reply:
Addressed in d99098c